### ISSUE TITLE
getPropertyName does not convert hyphenated attribute names to camelCased property names

### ISSUE THREAD

**Issue by amarajs** [hybridsjs/hybrids#8]
Hi @smalluban, thanks again for creating a great library. Ran across the following bug recently:

The `getPropertyName` method is passed a string portion of a template literal. This can include bound attribute names _that may be hyphenated_. Here is the code that passes a template literal piece to `getPropertyName`:

```javascript
// html/template.js:324
if (value === PLACEHOLDER) {
  var propertyName = getPropertyName(rawParts[parts.length]);
  parts.push([compileIndex, resolveProperty(name, propertyName, isSVG)]);
```
However, `getPropertyName` does _not_ convert hyphenated attribute names to camelCase property names:

```javascript
// html/template.js:258
// if string = '" start-count="' then the return
// value is 'start-count'; it should be 'startCount'
function getPropertyName(string) {
  return string.replace(/\s*=\s*['"]*$/g, '').split(' ').pop();
}
```
This causes a bug because the first check `resolveProperty` makes is to see if the passed `propertyName` is _in_ the target element. But since the hyphenated property does not exist in the target, it ends up using the default value.

```javascript
// html/template.js:155
if (!isSVG && propertyName in target) {
  if (target[propertyName] !== value) {
    target[propertyName] = value;
```

If `getPropertyName` returned a camelCased name, I think everything would work as expected.

[Here is sample code](https://stackblitz.com/edit/hybrids-simple-counter-kc24t2?file=index.html) that reproduces the issue.

**Comment by smalluban** [comment]
💥@amarajs Thank you for submitting an issue with very well explained code example! 

### EDIT

I wrote my explanation below and started to think, that something more is going on here. Property factory uses attribute only once when it is *connected* to DOM in `connect` callback (called by `connectedCallback`). In that time in your example there is no `start-count` attribute yet - it will be created in next `requestAnimationFrame` - this is how `render` works - so in that time, your custom element will be already after connecting, and it will not use this attribute at all. Attributes should be used only as static values for your hybrids custom elements:

```javascript
html`
<simple-counter ... start-count="5"></simple-counter>
`;
```
If you want to pass dynamic values, you should use property name - `startCount`. It works like that by the design - attributes as static fallbacks and properties for dynamic changes. It is taken from `input` element implementation - attribute `value` is used only once. If you want to change value, you have to use property. Changing attribute has no effect.

Below longer explanation about attributes and properties :)

## Explanation

I read yesterday [RFC](https://github.com/robdodson/rfcs/blob/2a9e29eb0070555df68080ee2247df214eeee45e/text/0000-create-custom-element-type.md) of how react should support attributes in custom elements, and in one sentence there was an argument for not transforming attributes between camelCase and dashCase. I think now, that in some way they might be right... However, what you explained is not a bug. It should work like this. I will explain to you way (I have add it also to the readme in the property section for templating).

### Property factory

This part of the library automatically takes attribute value and use it once if a created element has one (in a dashed version of your property name). The main reason for that is to support cases, where the user is using your custom element directly in html document:

```html
<body>
   <my-element long-property-name="default value"></my-element>
</body>
```
In that case, your custom element has to support attributes in some way. There have to be some mapping between attributes and properties. In hybrids, property factory transforms property name to a dashed version for the attribute value. Why?

In mentioned RFC, they argue that usually transition between attributes and properties is done by lowercasing characters, for example, `tabindex` attribute maps to `tabIndex` property name. Yes - this is total truth, but there is yet another problem. What if our beloved browser vendors add to the element specification attribute with the same name as your custom element uses? What I can see, built-in elements have no dash in attribute names, the same like they have no dash in the name. I know, that adding dash to longer property names does not solve that problem for all cases, but it helps a lot.

### Templating

Creating templates with `html` and `render` factory is another story. You are in charge and you have all control over what and how you want to pass to custom elements in your template body. If you create an expression in an attribute, `html` is actually doing almost the same thing as property factory - it uses properties in favor of attributes as a fallback. If the author of the library requires using attribute (does not support corresponding property) - you can use it. If the author requires some strange named property - you can use it. And finally, If you are using custom elements defined by the hybrids (with properties defined by property factory) you should always use camelCased property name because they are always there. Only care about attributes as your public API for external users - that just want to add an element to HTML and use it.

```javascript
const MyElement = {
  render: ({ value }) => html`
    <some-external-element super-value="${value}></some-external-element>

    <my-internal-element superValue="${value}"></my-internal-element>
  `,
};
```

(EDIT: Like I wrote on the top, `super-value` attribute is set in RAF, not when the element is connected)

What is the most important in this case, if there would be a transformation from a dash to a camel in `getPropertyName` It would be even more confusing. It would be impossible to set `my-super-name` *property* name. It would be impossible to set `onmy-super-event` event callback. All because of transform to `mySuperName` and `onmySuperEvent`. 

To sum up, If you are using public `property` API, you can set dashed attribute and camel-cased property. In templates produces by `html` always try to use properties over attributes.

**Comment by amarajs** [comment]
Thanks for the great explanation. It will feel a little weird writing camelCased attributes in my HTML template but now that I look, it's [right here](https://github.com/hybridsjs/hybrids#properties--attributes) in the README. Thanks again for a great library!

**Comment by smalluban** [comment]
As it can be a little bit confusing, I will write a word about that case in the readme, so thanks for creating issue :)
